schemas/input/csl-data.json

```diff
@@ -472,24 +477,8 @@
       "anyOf": [
         {
           "properties": {
-            "date-parts": {
-              "type": "array",
-              "items": {
-                "type": "array",
-                "items": {
-                  "type": ["string", "number"]
-                },
-                "minItems": 1,
-                "maxItems": 3
-              },
-              "minItems": 1,
-              "maxItems": 2
-            },
-            "season": {
-              "type": ["string", "number"]
-            },
-            "circa": {
-              "type": ["string", "number", "boolean"]
```

Comment:
So this PR will remove `date-parts` entirely? Since this is a backwards incompatible change, it should be part of a major not minor release.

Another option is to note that `date-parts` is deprecated and will be removed in the next major release. This would mean for the rest of the 1.x releases, `date-parts` is still an options as well as `edtf`.

Reply:
Yeah, whatever you all think is fine. 

I'd particularly like to hear from people maintaining existing code: @fbennett, etc.

I partly removed it because I don't really understand the nuances of json schema like I do with rnc. I struggle enough with the basics without adding more complexity.

But we can revise it to allow that if there's consensus to that end.

Reply:
I think it is reasonable to maintain support for each of these formats, so long as we establish an order of priority:
1. edtf
2. date-parts
3. pandoc-style individual elements
4. raw
5. literal

Reply:
The technical issue I was struggling with, here and on the discussion about adding relations, is how I can say that a property's value may either be a string, or an object/array. I wasn't even clear it was possible.

But this works as expected, so it it.

``` json
{
  "description": " test schema validating choice of string or object",
  "type": "object",
  "anyOf": [
    {
      "properties": {
        "date": {
          "type": "string"
        }
      },
      "required": [
        "date"
      ]
    },
    {
      "properties": {
        "date": {
          "year": {
            "type": "number"
          }
        }
      },
      "additionalProperties": false
    }
  ]
}
```

So the question is, would this approach be acceptable, @dhimmel?

We'd still need more feedback if it's wise, but it seems the cleanest solution longer term.

Reply:
> I think it is reasonable to maintain support for each of these formats, so long as we establish an order of priority:
> 
> 1. edtf
> 2. date-parts
> 3. pandoc-style individual elements
> 4. raw
> 5. literal

We still need to clarify if we need the latter going forward. It's be cleaner to just say one representation or the other, but deprecate the older longer term.

Reply:
We already support both strings (literal, raw) and arrays (date parts), so I think validation is not an issue?

Reply:
It is if we want this to be valid:

``` yaml
issued: 2018-03-02
```

Reply:
Having date-variables like `issued` support a string or an object is doable using [`anyOf`](https://json-schema.org/understanding-json-schema/reference/combining.html) as @bdarcus notes above.

There is the unresolved syntax question of whether we should support the following options:

```yaml
- id: my-reference
  type: report
  # option 1: flat
  issued: 2020-06-29
  # option 2: nested
  issued:
    edtf: 2020-06-29
    date-parts: [[2020, 6, 29]]
```

I think the best longterm option is 1, iff we are 100% sure EDTF is the best date-serialization format that will ever exist for bibliographic records and will suit this project's needs forever.

Supporting option 2 has the benefit that you can create CSL JSON that old and new citeprocs can handle. It allows for more backwards compatibility.

We could support option 1 and option 2. This is what I'm currently leaning towards

> We still need to clarify if we need the latter going forward. It'd be cleaner to just say one representation or the other, but deprecate the older longer term.

I think we should note that edtf format is "preferred" immediately. In the longterm, we should consider deprecating the older formats if we feel there is no longer any use for them. Removing the alternative formats helps simplify the burden on citeproc implementations, so its something to keep in mind for future MAJOR releases.

Reply:
Maintaining backwards compatibility is important. There is a lot of embedded CSL-JSON out there that needs to remain valid. I think we should add edtf support now and specify the parsing order of priority as I gave above (e.g., in @dhimmel's second example, the date-parts will be ignored entirely in favor of the edtf).

Reply:
> We could support option 1 and option 2. This is what I'm currently leaning towards.

Me too.

It could end up that 2 is more common in the json, and 1 more common in the yaml. And I think that's fine, given their different purposes.

Reply:
So the processor instructions for (1) will be to treat this as a `raw` string that should be parsed first for edtf, then treated as literal if that fails, maybe following a process similar to @retorquere's below?

Reply:
If this is the direction we want to go, I'll revise the PR to at least add back the earlier code, and do what I can on the json schema details.

Reply:
So is the proposal then:
1. date-variables can be either a string or an object
1. if it's an object, it can have one or more of these named elements:
  1. elements
    1. edtf
    1. date-parts
    1. {pandoc-style individual elements}
       1. year
       1. month
       1. season
       1. day
    1. raw
    1. literal
    1. circa
  1. Elements are used in the listed order of priority (e.g., date-parts is ignored if edtf is present; literal is ignored if raw is present)
    1. circa is an exception. It is a supplement to the other non-edtf elements elements indicating an uncertain date)
  1. if `raw` is supplied, it is parsed as edft (and for other date formats?) following a procedure like Emiliano's
1. If it's a string, it is treated as `raw` and parsed the same way
1. Using EDTF-formatted dates entered in either string format or in the edtf element of an object is strongly preferred.

Reply:
> It could end up that 2 is more common in the json, and 1 more common in the yaml. And I think that's fine, given their different priorities.

it seems like there is a large preference for avoiding nested values, so for consistency with this design principle, I like support option 1 (in addition to 2 for compat).

> So the processor instructions for (1) will be to treat this as a raw string that should be parsed first for edtf, then treated as literal if that fails, maybe following a process similar to @retorquere's below?

I like the idea of try edtf, and fallback to literal if parsing fails. I don't think it's a good idea to add all the additional fallbacks that @retorquere's implementation currently performs. Implementing all of these fallbacks in the same way across citeproc implementations in different languages would be a big burden. I think we want consistent behavior across citeproc implementations, which becomes unlikely if the spec gets too complex.

Reply:
> So is the proposal then ...

Those details are TBD, but I'd rather not support the pandoc modeling here,
if we can avoid it.

I suggested EDTF as a better concise option for YAML.

It just adds more complexity all around to support three different
models/syntaxes.

But if @jgm objects, of course we can consider.

Part of the challenge is really just my limitations, and the limitations of
json schema.

With rnc, doing basically anything we want is easy and intuitive for me.

With json schema, it is not; at all.

So the more complex things get, the more I struggle to implement it.

But I guess that justifies narrative discussion before implementation.

We need input on this question from the people writing and maintaining the
CSL processors.

Reply:
re: pandoc parts, considering that only pandoc has currently adopted that syntax, so long as John is amenable to the edtf string structure, we could always not include it in the schema and pandoc could continue to support it as legacy off-spec behavior for compatibility.

Reply:
Before I commit this @dhimmel, does this look right, aside from my notes in the comment?

https://gist.github.com/bdarcus/30807ba6466b3a089afc2b62be68bbc7#file-edtf-version-json-L403

Reply:
Nevermind @dhimmel - just committed here.